lib/pbench/server/database/models/datasets.py

```diff
@@ -675,7 +650,7 @@ def as_dict(self) -> Dict[str, Any]:
             "access": self.access,
             "created": self.created.isoformat() if self.created else None,
             "name": self.name,
-            "owner": self.owner.username,
+            "owner": self.owner,
```

Comment:
OK: not having to make a call here to translate `owner_id` into a username may be an advantage, but again I'm worried about stale username strings if they're changeable. If not ... maybe this is enough to justify the second column as a cache.

Reply:
We definitely need a "UUID to username" and "username to UUID" service, which should be based on a Keycloak broker API. Iff we experience serious performance issues we could consider working out a way to manage a local cache, but I don't want to manage that cache and I'd much rather avoid it. (In any case, if we added it later it'd be transparent behind the translator service, which might be part of Nikhil's separate OIDC client.)

Reply:
Encapsulating the translations so that we have flexibility in implementing them sounds like an excellent idea.

My foremost concerns are that we should not be using "username" to identify a user (e.g., to identify the owner of a dataset) -- I think we've come to agreement on that -- and that we need to be very careful in our thinking about what we can do with a username.

The UUID-to-username translation is, I think, a solvable problem.  (Apparently we can do that for the requestor's UUID by looking in the token; for other UUIDs, we can either use cached values or we can go ask Keycloak.)  The username-to-UUID translation is totally unreliable -- it may produce multiple results, and the results that it produces may change from moment to moment.

So, frankly, I think we should limit ourselves to a one-way translation:
- we can tell the client that a particular asset is owned by a particular username (by translating the UUID)
- we can let the client search for assets related to a particular username (by translating the assets' UUIDs to usernames and then filtering the results)

But, we cannot do something like "add this user (by name) to this group" -- for that we need a UUID.  Now, we can try to help the user out by providing a search-for-UUIDs based on usernames, but for the actual add request, we can only accept a UUID.  And, we need to be on the lookout for other, analogous situations, because they may not be obvious at first blush.

Reply:
> But, we cannot do something like "add this user (by name) to this group"

This is why I want our captive Keycloak broker to manage groups and roles for us: it's the authority on managing "identity".

> we can let the client search for assets related to a particular username (by translating the assets' UUIDs to usernames and then filtering the results)

This implies storing stale usernames with the data (SQL and Elastic) so we can search on it, *or* ignoring all of the filtering capabilities of the engine and retrieving *all* results (regardless of name) so that we can filter by UUID reverse-translations after the fact. I don't find either option acceptable.

I think we should just query our Keycloak broker based on an input username. If two separate records are returned, meaning two separate users from external identity providers chose the same username, we'd need a mechanism to disambiguate.

So maybe we either just require a client who wants to filter by username to ask our broker, or we build our own translation API to encapsulate that operation... and build our APIs to take user UUID, just as we now identify datasets by resource_id rather than name. Similarly, if a client wants to "beautify" a dataset display by showing an owning username, they do it themselves from our UUID.

Reply:
Nikhil and I uncovered a couple more wrinkles.

In the currently proposed model, assuming that all token validation done by the Pbench Server uses off-line validation, the Pbench Server never actually connects to the Keycloak server.  I don't know whether that is something that we'll be able to maintain (especially once we implement groups and roles), but I think that that is kind of cool and nice.  However, it does mean that (if we maintain the separation) the Pbench Server cannot translate between UUIDs and usernames (unless it uses a caching strategy).

The other wrinkle is that the request to the Keycloak server to translate between username and UUID (in either direction) appears to be privileged.  There is a nice way to grant that privilege to the client (as opposed to any particular user), but it requires making the client "private", which requires configuring the Keycloak server to recognize the client and which requires providing the client with a "secret".  Aside from the fact that I would prefer that the Pbench Dashboard be a "public" client, the secret would presumably have to come from the Pbench Server via the API (I don't really want to build it into the client, and I don't think that doing so will protect it from leaking), which probably means that it will be available to _any_ client, which means it won't be particularly "secret".  Now, this, in and of itself, is not necessarily a problem, provided that we don't need to grant the client anything more than enhanced read privileges (i.e., enough to translate between UUIDs and usernames), but it's slightly unsatisfying.

The alternative is to make the Pbench Server into a private client of the Keycloak server.  This would prevent the secret from leaking, and it would enable the Pbench Server to do the UUID-username translations.  But, it would violate the nice separation that we have now.  Still, it would lay the foundation for other privileged operations, such as group and role management, which we might not want to entrust to the client or to an individual user.

Reply:
> Nikhil and I uncovered a couple more wrinkles.
> 
> In the currently proposed model, assuming that all token validation done by the Pbench Server uses off-line validation, the Pbench Server never actually connects to the Keycloak server. I don't know whether that is something that we'll be able to maintain (especially once we implement groups and roles), but I think that that is kind of cool and nice. However, it does mean that (if we maintain the separation) the Pbench Server cannot translate between UUIDs and usernames (unless it uses a caching strategy).

We need to be able to talk to the Keycloak broker at any time; we'll need to do it regularly when we add roles and groups. I think pretending this is unusual or impractical is a dead end.

> The other wrinkle is that the request to the Keycloak server to translate between username and UUID (in either direction) appears to be privileged. There is a nice way to grant that privilege to the client (as opposed to any particular user), but it requires making the client "private", which requires configuring the Keycloak server to recognize the client and which requires providing the client with a "secret". Aside from the fact that I would prefer that the Pbench Dashboard be a "public" client, the secret would presumably have to come from the Pbench Server via the API (I don't really want to build it into the client, and I don't think that doing so will protect it from leaking), which probably means that it will be available to _any_ client, which means it won't be particularly "secret". Now, this, in and of itself, is not necessarily a problem, provided that we don't need to grant the client anything more than enhanced read privileges (i.e., enough to translate between UUIDs and usernames), but it's slightly unsatisfying.

This is a big deal, definitely; although it makes sense. I can see Keycloak being willing to give any authenticated user their own profile info, but, yeah, being able to query anyone's UUID or username might be more questionable.

Even if we could manage passing on enhanced capabilities to a client, I agree that making the "secret" public sounds like a bad idea. But providing a Pbench Server API to do bidirectional translation within the scope of users is maybe a less flagrant wart? Maybe?

It might also argue for continuing to do username at our API boundaries. And outgoing that's no problem because the actual UUID that owns a dataset is always going to have a single unique username translation. Processing incoming usernames for search is a different matter.

One option is simply to not do that. For example, we could drop the ability to ask for datasets owned by a specific user, and only support "mine" and "notmine" with a boolean parameter. (Or only support "user_id" and how they get it is their own problem... though that's not entirely satisfying.)

> The alternative is to make the Pbench Server into a private client of the Keycloak server. This would prevent the secret from leaking, and it would enable the Pbench Server to do the UUID-username translations. But, it would violate the nice separation that we have now. Still, it would lay the foundation for other privileged operations, such as group and role management, which we might not want to entrust to the client or to an individual user.

I think we definitely will need to do this because we *are* going to want to manage roles and groups. Whether we want to provide services to disambiguate usernames is a subset of this question. Maybe we'll have to, but maybe we're better off deferring that question for now and going with something simple like my "mine/notmine" proposal above. This would change a couple of queries where we allow specifying an owner by name (e.g., `/datasets/list`).

Reply:
yeah and I think that is an important distinction between querying our own user table and making a request to Keycloak to give us info pertaining to a different user, we have to use some sort of a token that has enough privileges to obtain the information it is trying to access. This can be done by a service client token by granting it a "view_user" access, these are special tokens that don't have refresh tokens attached to them so when they expire they are gone which might be useful for our case. 

Another a little bit hacky way would be to bypass the Keycloak broker and query the postgres database that our Keycloak is using, we control the Keycloak broker and so we control the postgres DB that it uses so we can query it directly without needing any tokens. :stuck_out_tongue_closed_eyes:

Reply:
> little bit hacky way

Don't be seduced by the Dark Side!

Reply:
> > little bit hacky way
> 
> Don't be seduced by the Dark Side!

Indeed. And doubly so if Keycloak's SQL schema isn't published and supported for third party use ... ☠️